pkg/kubelet/envvars/envvars.go

```diff
@@ -56,33 +67,42 @@ func makeEnvVariableName(str string) string {
 	return strings.ToUpper(strings.Replace(str, "-", "_", -1))
 }
 
-func makeLinkVariables(service api.Service) []api.EnvVar {
+func makeLinkVariables(service *api.Service) []api.EnvVar {
 	prefix := makeEnvVariableName(service.Name)
-	protocol := string(api.ProtocolTCP)
-	if service.Spec.Protocol != "" {
-		protocol = string(service.Spec.Protocol)
-	}
-	portPrefix := fmt.Sprintf("%s_PORT_%d_%s", prefix, service.Spec.Port, strings.ToUpper(protocol))
-	return []api.EnvVar{
-		{
-			Name:  prefix + "_PORT",
-			Value: fmt.Sprintf("%s://%s:%d", strings.ToLower(protocol), service.Spec.PortalIP, service.Spec.Port),
-		},
-		{
-			Name:  portPrefix,
-			Value: fmt.Sprintf("%s://%s:%d", strings.ToLower(protocol), service.Spec.PortalIP, service.Spec.Port),
-		},
-		{
-			Name:  portPrefix + "_PROTO",
-			Value: strings.ToLower(protocol),
-		},
-		{
-			Name:  portPrefix + "_PORT",
-			Value: strconv.Itoa(service.Spec.Port),
-		},
-		{
-			Name:  portPrefix + "_ADDR",
-			Value: service.Spec.PortalIP,
-		},
+	all := []api.EnvVar{}
+	for i := range service.Spec.Ports {
+		sp := &service.Spec.Ports[i]
+
+		protocol := string(api.ProtocolTCP)
+		if sp.Protocol != "" {
+			protocol = string(sp.Protocol)
+		}
+		if i == 0 {
+			// Docker special-cases the first port.
+			all = append(all, api.EnvVar{
+				Name:  prefix + "_PORT",
+				Value: fmt.Sprintf("%s://%s:%d", strings.ToLower(protocol), service.Spec.PortalIP, sp.Port),
+			})
+		}
+		portPrefix := fmt.Sprintf("%s_PORT_%d_%s", prefix, sp.Port, strings.ToUpper(protocol))
```

Comment:
I tried to convince myself of that, but could not get docker to do anything
but the obvious thing.  But it should not matter for us - there is only the
well-known service port - there is no "real" port.

On Mon, Mar 30, 2015 at 4:45 PM, Angus Lees <notifications@github.com>
wrote:

> In pkg/kubelet/envvars/envvars.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6182#discussion_r27442718>
> :
>
> > +	all := []api.EnvVar{}
> > +	for i := range service.Spec.Ports {
> > +		sp := &service.Spec.Ports[i]
> > +
> > +		protocol := string(api.ProtocolTCP)
> > +		if sp.Protocol != "" {
> > +			protocol = string(sp.Protocol)
> > +		}
> > +		if i == 0 {
> > +			// Docker special-cases the first port.
> > +			all = append(all, api.EnvVar{
> > +				Name:  prefix + "_PORT",
> > +				Value: fmt.Sprintf("%s://%s:%d", strings.ToLower(protocol), service.Spec.PortalIP, sp.Port),
> > +			})
> > +		}
> > +		portPrefix := fmt.Sprintf("%s_PORT_%d_%s", prefix, sp.Port, strings.ToUpper(protocol))
>
> I'm no expert here, but my understanding was that this was always a bug
> with the link variables created by k8s. I believe the variable *names*
> should use TargetPort, and the variable *values* use Port. The idea is
> that the port numbers declared by containers are "well known", and you want
> to find out (eg) "where has mysql:3306 been made available in this
> particular setup?"
>
> I assumed it would be too disruptive to change the existing k8s behaviour
> and so never raised the bug - but perhaps it could be fixed as part of
> this? (Assuming others agree with my understanding of link variables)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6182/files#r27442718>
> .
>